### ISSUE TITLE
Speed optimizations

### ISSUE THREAD

**Issue by tbreloff** [tbreloff/Plots.jl#357]
There is much to improve in time-to-first-plot and time-to-nth plot.  I should investigate using SnoopCompile for better precompilation, and do some benchmarking of common plot types to figure out what parts are the worst offenders.

ref: #344

**Comment by timholy** [comment]
It may depend a lot on the backend. If it's Gadfly/Compose, then most of the precompile statements that help may [already be there](https://github.com/dcjones/Gadfly.jl/blob/master/src/precompile.jl) (though it's possible they might need updating). There's a limit to how much precompilation can help, because native code that involves interactions between modules is not cached, and of course in julia there's a lot of code that involves, e.g., types defined in a package and methods defined in Base.

For other backends, this would of course be a worthy experiment. If you want to experiment with Gadfly too, note the existence of a script to [regenerate the precompiles](https://github.com/dcjones/Gadfly.jl/blob/master/test/regen-precompiles.jl).

**Comment by tbreloff** [comment]
Thanks for the links @timholy.  For improvements to time-to-first-plot, I think there's a lot of Plots-only precompilation that would help, but some things have no foreseeable solution (like the python setup code inside PyPlot).  I might make a first pass at this, but I'm sure someone more knowledgeable about julia compilation can help more than me.

Also as we discussed at JuliaCon, I want to nail down whether I can cut down on object copies (either new logic, or a copy-on-write type), some of which could benefit from a few redesigns.  I won't tackle this stuff immediately, and I have a lot of benchmarking to do before any decisions are made.

**Comment by timholy** [comment]
I should have also mentioned: Gadfly's precompiles were generated by running the test suite; so if there's no heatmap in the testsuite, that could explain some of the slowness. Also, precompilation will help if users pass the same kind of data, but precompiles for `Float64` won't help if the user passes an array of `Float32` instead...

**Comment by tbreloff** [comment]
You can ignore the heatmap issue for now... I don't think that's related to compilation... I think it's making different PyPlot calls.

I followed your instructions in SnoopCompile.  For this simple script:

```julia
using Plots
plot(rand(10))
png("/tmp/tmp")
```

Before:

```
tom@tom-office-ubuntu:~/.julia/v0.4/PlotRecipes$ time julia /tmp/testplots.jl 
[Plots.jl] Initializing backend: pyplot

real	0m21.941s
user	0m21.572s
sys	0m0.644s
```

after:

```
tom@tom-office-ubuntu:~/.julia/v0.4/PlotRecipes$ time julia /tmp/testplots.jl 
[Plots.jl] Initializing backend: pyplot

real	0m18.031s
user	0m17.632s
sys	0m0.648s
tom@tom-office-ubuntu:~/.julia/v0.4/PlotRecipes$ 
```

Here's [my snoop script](https://github.com/tbreloff/Plots.jl/blob/dev/test/snoop.jl) and [my generated file](https://github.com/tbreloff/Plots.jl/blob/dev/src/precompile.jl)... I'll need to do more benchmarking to figure out where all the time is spent.  GR, for example, is twice as fast, even though I didn't run any GR examples during snooping.

@timholy Don't feel obligated to spend any time on this... I'm more posting it as a reference for myself.

**Comment by jebej** [comment]
Plots is still really slow compared to using PyPlot directly:
```julia
julia> @time (using Plots; plot([1,2,3],[1,2,3]))
 19.371098 seconds (18.73 M allocations: 794.405 MB, 2.06% gc time)
```
and in an other julia session:
```julia
julia> @time (using PyPlot; plot([1,2,3],[1,2,3]))
  3.327282 seconds (2.42 M allocations: 104.145 MB, 4.92% gc time)
```

**Comment by mkborregaard** [comment]
That's true. Though the timings in your example are probably because precompilation is currently deactivated in Plots releases, at the request of the Metadata maintainers (though it generally works so we keep it activated on master).

**Comment by jebej** [comment]
Enabling precompilation on the current release shaved off ~6 secs. Interestingly it also changed the default backend to GR...

```julia
julia> @time (using Plots; pyplot(); plot([1,2,3],[1,2,3]))
 13.782701 seconds (16.71 M allocations: 702.408 MB, 2.66% gc time)
```

**Comment by mkborregaard** [comment]
Yes, we changed the first-choice backend to GR on the latest release (as of today) because of a number of issues with pyplot - one being that there is a world-age issue with pyplot that is resolved by manually calling `pyplot()`, rather than defaulting to it.
